exonum/src/blockchain/mod.rs

```diff
@@ -307,7 +305,31 @@ impl BlockchainMut {
                 call_errors.put(&location, error);
             }
         }
-        // Get tx & state hash.
+
+        let block = self.create_block_header(proposer_id, height, tx_hashes);
+        log::trace!("execute block = {:?}", block);
+
+        // Calculate block hash.
+        let block_hash = block.object_hash();
+        // Update height.
+        let schema = Schema::new(&fork);
+        schema.block_hashes_by_height().push(block_hash);
+        // Save block.
+        schema.blocks().put(&block_hash, block);
+        (block_hash, fork.into_patch())
+    }
+
+    fn create_block_header(
+        &self,
+        proposer_id: ValidatorId,
```

Comment:
Slightly off-topic: Some time ago there were talks about replacing `ValidatorId` with its `PublicKey` (since `ValidatorId` not always directly maps to validators if the configuration changes, see [example](https://github.com/exonum/exonum/blob/master/services/supervisor/tests/supervisor/consensus_config.rs#L127)).
Since we didn't come to agreement at this point, maybe make the `proposer_id` an optional entry too?

cc @slowli @alekseysidorov @aleksuss

Reply:
Sort of makes sense. AFAIK, presently `ValidatorId` field is used in the request processing algorithm (if a node has received a block, it may ask the proposer about transactions in the block). Since we plan to update network logic, making `proposer_id` optional seems logical.

Reply:
So, if I understand correctly, I need to move `proposer_id` to block header entries?

Reply:
Yes (be sure to document it this as a breaking change).